### PR TITLE
adds a file/:hash endpoint for file data over api

### PR DESCRIPTION
**Pull request by andrewxhill**
not certain if this is the right home, happy to update if you have thoughts

files/:block conflicted with the obvious place for it, so created a file/:hash api => fileData method

this solved my issues of needing to get the raw data from the api though

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: f3645d4bac1418201c6369e4e8fe9c4611059626
  * Commits must be in the following format: **%{scope}: %{description}**



---
This message was auto-generated by https://gitcop.com

**Comment by andrewxhill** [comment]
if this gets the 👍 as the right way to go, i'd like to extend it to include an endpoint for `blockData`. I'm thinking it would work like, `blockData(<blockId>, <fileIndex>)` this would allow the client to skip back-n-forth parsing steps a lot of the time

**Comment by sanderpick** [comment]
Based on the convo from this morning, I think the two handlers you want here are the ones described in this comment @andrewxhill: https://github.com/textileio/go-textile/issues/688#issuecomment-487094089. The endpoints would take the _path_ to the file within the DAG as described in the issue.

No need to holdup by the gitcop, but please tag on a subject to future commits. It helps to get an idea about what a chunk of commits were all about.

File: core/api_files.go [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-	"strconv"
-
 	"github.com/gin-gonic/gin"
 	ipld "github.com/ipfs/go-ipld-format"
 	"github.com/textileio/go-textile/ipfs"
 	"github.com/textileio/go-textile/pb"
 	"github.com/textileio/go-textile/schema"
+	"io/ioutil"
+	"net/http"
+	"strconv"
```
Comment (sanderpick) [inline core/api_files.go]: There a `make fmt` task that should clean these up

File: core/api_files.go [hunk 2]
```diff
@@ -190,3 +190,33 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
 
 	pbJSON(g, http.StatusOK, keys)
 }
+
+// lsThreadFileTargetKeys godoc
```
Comment (sanderpick) [inline core/api_files.go]: these generate the swagger docs, so would need to be updated and then run `make docs`

File: core/api_files.go [hunk 3]
```diff
@@ -190,3 +190,33 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
...
+	}
+
+	reader.Seek(0, 0)
+
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		g.String(http.StatusBadRequest, err.Error())
+	}
+
+	contentType := http.DetectContentType(data)
```
Comment (sanderpick) [inline core/api_files.go]: I think  you can do this faster by first getting the file index (what we're going to call file meta data) object, which already has the mime type and content length of the underlying data. Then you can use `g.DataFromReader`, which means that you don't need to read the entire data into memory before serving it back.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: f3645d4bac1418201c6369e4e8fe9c4611059626
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f15c3830500fec5260199a489cd8c613c73b3dd4
  * Commits must be in the following format: **%{scope}: %{description}**



---
This message was auto-generated by https://gitcop.com

**Comment by andrewxhill** [comment]
Okay, so leaving the block based stuff out of the PR (in favor of https://github.com/textileio/go-textile/issues/688#issuecomment-487094089 work), the getFileData method is gtg and means we can make the js lib update asap

File: core/api_files.go [hunk 4]
```diff
@@ -190,3 +190,35 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
...
+
+// getFileData godoc
+// @Summary File data at hash
+// @Description Returns raw data for file
+// @Tags files
+// @Produce application/octet-stream
+// @Param hash path string true "file hash"
+// @Success 200 {string} byte
+// @Failure 400 {string} string "Bad Request"
+// @Router /files/{hash} [get]
```
Comment (sanderpick) [inline core/api_files.go]: file or files? Based on the implementation, seems like this should be `/file/:hash/data`.

File: core/api_files.go [hunk 5]
```diff
@@ -190,3 +190,35 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
...
+func (a *api) getFileData(g *gin.Context) {
+	hash := g.Param("hash")
+
+	var err error
```
Comment (sanderpick) [inline core/api_files.go]: not needed

File: core/api_files.go [hunk 6]
```diff
@@ -190,3 +190,35 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
...
+
+	file, err := a.node.FileIndex(hash)
+	if err != nil {
+		g.String(http.StatusBadRequest, err.Error())
```
Comment (sanderpick) [inline core/api_files.go]: needs a `return` on the next line

Comment (sanderpick) [reply core/api_files.go]: also, should be 404

File: core/api_files.go [hunk 7]
```diff
@@ -190,3 +190,35 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
...
+	}
+
+	contentLength := file.GetSize()
+	contentType := file.GetMedia()
+
+	reader, _, err := a.node.FileData(hash)
+	if err != nil {
+		g.String(http.StatusBadRequest, err.Error())
```
Comment (sanderpick) [inline core/api_files.go]: same here, 404 followed by `return`

File: core/api_files.go [hunk 8]
```diff
@@ -190,3 +190,35 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
...
+	}
+
+	reader.Seek(0, 0)
```
Comment (sanderpick) [inline core/api_files.go]: not needed as you haven't read anything off the reader yet

File: core/api_files.go (hunk 4 above)
Comment (andrewxhill) [reply core/api_files.go]: ah yep, sloppy. all fixed

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: f3645d4bac1418201c6369e4e8fe9c4611059626
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f15c3830500fec5260199a489cd8c613c73b3dd4
  * Commits must be in the following format: **%{scope}: %{description}**



---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: f3645d4bac1418201c6369e4e8fe9c4611059626
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f15c3830500fec5260199a489cd8c613c73b3dd4
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f23e145ea62172457e0a2eff4524b8c0c633be73
  * Commits must be in the following format: **%{scope}: %{description}**



---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: f3645d4bac1418201c6369e4e8fe9c4611059626
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f15c3830500fec5260199a489cd8c613c73b3dd4
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f23e145ea62172457e0a2eff4524b8c0c633be73
  * Commits must be in the following format: **%{scope}: %{description}**



---
This message was auto-generated by https://gitcop.com

File: core/api_files.go [hunk 9]
```diff
@@ -190,3 +190,33 @@ func (a *api) lsThreadFileTargetKeys(g *gin.Context) {
...
+	file, err := a.node.FileIndex(hash)
+	if err != nil {
+		g.String(http.StatusNotFound, err.Error())
+		return
+	}
+
+	contentLength := file.GetSize()
+	contentType := file.GetMedia()
+
+	reader, _, err := a.node.FileData(hash)
```
Comment (sanderpick) [inline core/api_files.go]: doh, I just remembered this method returns the file index as it second param, may as well use it instead of loading it twice.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: f3645d4bac1418201c6369e4e8fe9c4611059626
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f15c3830500fec5260199a489cd8c613c73b3dd4
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: f23e145ea62172457e0a2eff4524b8c0c633be73
  * Commits must be in the following format: **%{scope}: %{description}**



---
This message was auto-generated by https://gitcop.com

**Comment by balupton** [comment]
Tried this locally and all I get is 404's... any ideas?

**Comment by andrewxhill** [comment]
Hmm.. give me a few more details. I'm running it currently through my local desktop install

**Comment by balupton** [comment]
So I run the daemon via `textile daemon`:

```
> textile daemon
go-textile version: v0.1.12-rc3-1-g3bb4ad52
Repo version: 13
Repo path: /Users/balupton/.textile/repo
API address: 127.0.0.1:40600
Gateway address: 127.0.0.1:5050
System version: amd64/darwin
Golang version: go1.12.4
PeerID:  12D3KooWJoRPGDJMHziS7V1JoSxq4LfQ4kajNqqiqUfATSFsMJqC
Account: P6XdG9bpvBqEdXsZsNX25HTuXjEatSndRaSB79Lx65bacEZ7
```

The type `textile files ls` to get this block:

```
       {
            "block": "QmRmeF6xa6uoQGQG8U2mFUrqcmxR7Y2WY7gRs2ey4cgq1Z",
            "target": "QmRGJsjLaEp7YNzsQ6dwaVwBn4b73CHbvCLcszWKVNuRNV",
            "date": "2019-04-25T03:43:00.663899Z",
            "user": {
                "address": "P6XdG9bpvBqEdXsZsNX25HTuXjEatSndRaSB79Lx65bacEZ7",
                "name": "balupton-dev"
            },
            "files": [
                {
                    "links": {
                        "large": {
                            "mill": "/image/resize",
                            "checksum": "DUn51aKz1n4Ts2TEj12n6WgmXgjXp8SSU5LyDk918bqQ",
                            "source": "ATnpNZcdKLPBJYbpS7rsZZc3o2g8PywsNaZbjFvCPbVa",
                            "opts": "21uBAuSeQUdw5aDu5CYPxEfeiLVeuvku1T26nWtJC84C",
                            "hash": "QmS6fNpJKswtRizEuTKYPSNNG8jm2hpNxKwtnm9nPY4fNz",
                            "key": "7pyCYJe7YRgCw86Bn2URD6tU1bnnhyNwp6HgfgeRNaz2ybFVq4fszCxVfRj3",
                            "media": "image/jpeg",
                            "name": "stdin",
                            "size": "23275",
                            "added": "2019-04-25T03:42:59.906707Z",
                            "meta": {
                                    "height": 425,
                                    "width": 425
                                },
                            "targets": [
                                "QmRGJsjLaEp7YNzsQ6dwaVwBn4b73CHbvCLcszWKVNuRNV"
                            ]
                        },
                        "small": {
                            "mill": "/image/resize",
                            "checksum": "9MogY5zLtXx3exfHbusJJX34uBuKuW322UarWenU1Mk7",
                            "source": "ATnpNZcdKLPBJYbpS7rsZZc3o2g8PywsNaZbjFvCPbVa",
                            "opts": "CassDcqf192MnceweJKGJvhZfrV9kB3GJRbvNPWm6raa",
                            "hash": "QmVjMc5xfEcDD2w7eCo7Mz8YcD744JNks6yRz9T3gbirWx",
                            "key": "HoN5F6bXi6YBgRfNpNT8bSvMGfyM3sQ5bfoE4SL76wMAdBpnyU7Q2axFb8Rr",
                            "media": "image/jpeg",
                            "name": "stdin",
                            "size": "14478",
                            "added": "2019-04-25T03:43:00.000646Z",
                            "meta": {
                                    "height": 320,
                                    "width": 320
                                },
                            "targets": [
                                "QmRGJsjLaEp7YNzsQ6dwaVwBn4b73CHbvCLcszWKVNuRNV"
                            ]
                        },
                        "thumb": {
                            "mill": "/image/resize",
                            "checksum": "6Nf3DcZemisdztipeq9CRAhFHHUSFoTEeeBdP2x3d55h",
                            "source": "DUn51aKz1n4Ts2TEj12n6WgmXgjXp8SSU5LyDk918bqQ",
                            "opts": "7aGVJ7nGgmHdqv8oiEKZ2ZbrNBv1zVP3ADSuT2sW3MwT",
                            "hash": "Qmeky4j48GsjSa6Y6jU2QjaRkNRAWoKyxcRLyxRKjp3uqw",
                            "key": "2AuyoU4h5Kgzs8r7TJSxJqGHrEwkfRFt2eSSroz2YNT8sK5rhhkxPVeXZMgVZ",
                            "media": "image/jpeg",
                            "name": "stdin",
                            "size": "3030",
                            "added": "2019-04-25T03:43:00.070141Z",
                            "meta": {
                                    "height": 100,
                                    "width": 100
                                },
                            "targets": [
                                "QmRGJsjLaEp7YNzsQ6dwaVwBn4b73CHbvCLcszWKVNuRNV"
                            ]
                        }
                    }
                }
            ],
            "comments": [
            ],
            "likes": [
            ],
            "threads": [
                "12D3KooWQMJcNcbws6fz9aBNVpqLYGmfsJVphfKWD6i8Zxd1ZsAP"
            ]
        }
    ]
}
```

All these URLs give 404s.

These ones give 404 with white page:

- http://127.0.0.1:40600/file/QmS6fNpJKswtRizEuTKYPSNNG8jm2hpNxKwtnm9nPY4fNz/data
- http://127.0.0.1:40600/file/QmRmeF6xa6uoQGQG8U2mFUrqcmxR7Y2WY7gRs2ey4cgq1Z/data
- http://127.0.0.1:40600/file/QmRGJsjLaEp7YNzsQ6dwaVwBn4b73CHbvCLcszWKVNuRNV/data
- http://127.0.0.1:40600/file/ATnpNZcdKLPBJYbpS7rsZZc3o2g8PywsNaZbjFvCPbVa/data

These give 404 with black page:

- http://127.0.0.1:5050/file/QmS6fNpJKswtRizEuTKYPSNNG8jm2hpNxKwtnm9nPY4fNz/data
- http://127.0.0.1:5050/file/QmRmeF6xa6uoQGQG8U2mFUrqcmxR7Y2WY7gRs2ey4cgq1Z/data
- http://127.0.0.1:5050/file/QmRGJsjLaEp7YNzsQ6dwaVwBn4b73CHbvCLcszWKVNuRNV/data
- http://127.0.0.1:5050/file/ATnpNZcdKLPBJYbpS7rsZZc3o2g8PywsNaZbjFvCPbVa/data

**Pull request merged**